### ISSUE TITLE
Sending a message with "Mention" (in fbchat.models)

### ISSUE THREAD

**Issue by nguyenvulong** [carpedm20/fbchat#200]
I want to send a message to a group chat, and in that message - I want to mentioned a member of that group. 
But I don't really understand how to integrate fbchat.models with Client class. In other words, how can I send a message, for example, using "client.sendRemoteImage" in Client class , and mention (using Mention class) someone in that message?

Thank you

**Comment by madsmtm** [comment]
This is not currently possible, but it shouldn't be too hard to implement. Once we've fixed the crashing issues, and made `fbchat` more stable, I might get around to this

**Comment by svaikstude** [comment]
`def sendMessage(self, message, thread_id=None, 
                        thread_type=ThreadType.USER, **kwargs):
            """
            Sends a message to a thread

            :param message: Message to send
            :param thread_id: User/Group ID to send to. See :ref:`intro_threads`
            :param thread_type: See :ref:`intro_threads`
            :type thread_type: models.ThreadType
            :kwargs for @mention purpose {userID : (start, end)},
             where start is relative start position of @mention in a message
             and end is relative end position of @mention
            :return: :ref:`Message ID <intro_message_ids>` of the sent message
            :raises: FBchatException if request failed
            """
            thread_id, thread_type = self._getThread(thread_id, thread_type)
            data = self._getSendData(thread_id, thread_type)

            data['action_type'] = 'ma-type:user-generated-message'
            data['body'] = message or ''
            if kwargs:
                n = 0
                for key, value in kwargs.items():
                    data['profile_xmd[%d][id]'%n] = key
                    data['profile_xmd[%d][offset]'%n] = value[0]
                    data['profile_xmd[%d][length]'%n] = value[1] - value[0]
                    data['profile_xmd[%d][type]'%n] = 'p'
                    n += 1
            data['has_attachment'] = False
            data['specific_to_list[0]'] = 'fbid:' + thread_id
            data['specific_to_list[1]'] = 'fbid:' + self.uid

            return self._doSendRequest(data)`

This is my fix for @mention

**Comment by svaikstude** [comment]
`def sendMessage(self, message, thread_id=None, 
                thread_type=ThreadType.USER, **kwargs):
    """
    Sends a message to a thread

    :param message: Message to send
    :param thread_id: User/Group ID to send to. See :ref:`intro_threads`
    :param thread_type: See :ref:`intro_threads`
    :type thread_type: models.ThreadType
    :kwargs for @mention purpose {userID : (start, end)},
     where start is relative start position of @mention in a message
     and end is relative end position of @mention
    :return: :ref:`Message ID <intro_message_ids>` of the sent message
    :raises: FBchatException if request failed
    """
    thread_id, thread_type = self._getThread(thread_id, thread_type)
    data = self._getSendData(thread_id, thread_type)

    data['action_type'] = 'ma-type:user-generated-message'
    data['body'] = message or ''
    if kwargs:
        n = 0
        for key, value in kwargs.items():
            data['profile_xmd[%d][id]'%n] = key
            data['profile_xmd[%d][offset]'%n] = value[0]
            data['profile_xmd[%d][length]'%n] = value[1] - value[0]
            data['profile_xmd[%d][type]'%n] = 'p'
            n += 1
    data['has_attachment'] = False
    data['specific_to_list[0]'] = 'fbid:' + thread_id
    data['specific_to_list[1]'] = 'fbid:' + self.uid

    return self._doSendRequest(data)`

**Comment by svaikstude** [comment]
This is how I fixed @mention issue:
`def sendMessage(self, message, thread_id=None, 
                thread_type=ThreadType.USER, **kwargs):
    """
    Sends a message to a thread

    :param message: Message to send
    :param thread_id: User/Group ID to send to. See :ref:`intro_threads`
    :param thread_type: See :ref:`intro_threads`
    :type thread_type: models.ThreadType
    :kwargs for @mention purpose {userID : (start, end)},
     where start is relative start position of @mention in a message
     and end is relative end position of @mention
    :return: :ref:`Message ID <intro_message_ids>` of the sent message
    :raises: FBchatException if request failed
    """
    thread_id, thread_type = self._getThread(thread_id, thread_type)
    data = self._getSendData(thread_id, thread_type)

    data['action_type'] = 'ma-type:user-generated-message'
    data['body'] = message or ''
    if kwargs:
        n = 0
        for key, value in kwargs.items():
            data['profile_xmd[%d][id]'%n] = key
            data['profile_xmd[%d][offset]'%n] = value[0]
            data['profile_xmd[%d][length]'%n] = value[1] - value[0]
            data['profile_xmd[%d][type]'%n] = 'p'
            n += 1
    data['has_attachment'] = False
    data['specific_to_list[0]'] = 'fbid:' + thread_id
    data['specific_to_list[1]'] = 'fbid:' + self.uid

    return self._doSendRequest(data)`

**Comment by svaikstude** [comment]
`def sendMessage(self, message, mention=None, thread_id=None, 
                thread_type=ThreadType.USER):
    """
    Sends a message to a thread

    :param message: Message to send
    :param thread_id: User/Group ID to send to. See :ref:`intro_threads`
    :param thread_type: See :ref:`intro_threads`
    :type thread_type: models.ThreadType
    :mention is in this format {userID : (start, end)},
     where start is relative start position of @mention in a message
     and end is relative end position of @mention
    :return: :ref:`Message ID <intro_message_ids>` of the sent message
    :raises: FBchatException if request failed
    """
    thread_id, thread_type = self._getThread(thread_id, thread_type)
    data = self._getSendData(thread_id, thread_type)

    data['action_type'] = 'ma-type:user-generated-message'
    data['body'] = message or ''
    if mention:
        n = 0
        for key, value in mention.items():
            data['profile_xmd[%d][id]'%n] = key
            data['profile_xmd[%d][offset]'%n] = value[0]
            data['profile_xmd[%d][length]'%n] = value[1] - value[0]
            data['profile_xmd[%d][type]'%n] = 'p'
            n += 1
    data['has_attachment'] = False
    data['specific_to_list[0]'] = 'fbid:' + thread_id
    data['specific_to_list[1]'] = 'fbid:' + self.uid

    return self._doSendRequest(data)`


Hi, @nguyenvulong this is how I fixed @mention problem, hope this helps :)

**Comment by madsmtm** [comment]
This is now implemented in `1.1.0`, partial credit goes to `svaikstude`. You can now do something like this:
```python
client.sendLocalImage(image_local_url, Message(text='A @mention', mentions=[Mention('<thread_id>', offset=2, length=8)]))
```
